### PR TITLE
289: correctness metrics and tighter tests

### PR DESCRIPTION
**Pull request by maxkazmsft**

### CODE REVIEW

File: examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb [hunk 1]
```diff
@@ -747,28 +726,52 @@
...
+    "\n",
+    "# Tensorboard and Logging:\n",
+    "trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_training_output(summary_writer))\n",
+    "trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_validation_output(summary_writer))\n",
+    "\n",
+    "# add specific logger which also triggers printed metrics on test set\n",
+    "@trainer.on(Events.EPOCH_COMPLETED)\n",
+    "def log_training_results(engine):\n",
+    "    evaluator.run(train_loader)\n",
+    "    log_results(engine, evaluator, summary_writer, n_classes, stage=\"Training\")\n",
```
Comment (yalaudah) [inline examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb:759]: Can we have these two functions follow the same import pattern? Like either use:
```
tensorboard_handlers.log_results( ..
logging_handlers.log_metrics( ..
```
or 
```
log_results( ..
log_metrics( ..
```
i think it looks cleaner that way

File: examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb [hunk 2]
```diff
@@ -747,28 +726,52 @@
...
+    "    logging_handlers.log_metrics(engine, evaluator, stage=\"Training\")\n",
+    "\n",
+    "# add specific logger which also triggers printed metrics on validation set\n",
+    "@trainer.on(Events.EPOCH_COMPLETED)\n",
+    "def log_validation_results(engine):\n",
+    "    evaluator.run(val_loader)\n",
+    "    log_results(engine, evaluator, summary_writer, n_classes, stage=\"Validation\")\n",
+    "    logging_handlers.log_metrics(engine, evaluator, stage=\"Validation\")"
```
Comment (yalaudah) [inline examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb:767]: same comment here as above

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 3]
```diff
@@ -203,40 +206,43 @@ def run(*options, cfg=None, debug=False):
...
-            },
-            fname=fname,
-        ),
-    )
-
     # Tensorboard and Logging:
     trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_training_output(summary_writer))
     trainer.add_event_handler(Events.ITERATION_COMPLETED, tensorboard_handlers.log_validation_output(summary_writer))
 
+    # add specific logger which also triggers printed metrics on test set
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/train.py:217]: I'm not sure what you mean by this comment. Also, shouldn't it be "train" set not "test"?

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 4]
```diff
@@ -203,40 +206,43 @@ def run(*options, cfg=None, debug=False):
...
+
+        @trainer.on(Events.COMPLETED)
+        def log_final_validation_results(engine):
+            evaluator.run(val_loader)
+            metrics = evaluator.state.metrics
+            out_dict = {x: metrics[x] for x in ["nll", "pixacc", "mca", "mIoU"]}
+            with open(fname, "w") as fid:
+                json.dump(out_dict, fid)
+            log_msg = " ".join(f"{k}: {out_dict[k]}" for k in out_dict.keys())
+            logging.info(log_msg)
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/train.py:243]: the final validation results will still be logged right? so maybe we re-purpose this to just dump the final results to the disk.

also, is there a way we can avoid evaluating on the validation set twice for the last epoch? Ignite again is complicating this, but it would be nice if we can do it.

File: cv_lib/cv_lib/event_handlers/__init__.py [hunk 5]
```diff
@@ -32,7 +32,7 @@ def __call__(self, engine, to_save):
         self._checkpoint_handler(engine, to_save)
         if self._snapshot_function():
             files = glob.glob(os.path.join(self._model_save_location, self._running_model_prefix + "*"))
-            print(files)
+            # print(files)
```
[the remark this answers is not in the document]
Comment (maxkazmsft) [reply cv_lib/cv_lib/event_handlers/__init__.py:35]: Yup, done

File: examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb (hunk 1 above)
Comment (maxkazmsft) [reply examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb:759]: I always like specifying full library path - will go with the former

File: examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb (hunk 2 above)
Comment (maxkazmsft) [reply examples/interpretation/notebooks/Dutch_F3_patch_model_training_and_evaluation.ipynb:767]: also updated in train.py

File: experiments/interpretation/dutchf3_patch/local/train.py (hunk 3 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/train.py:217]: good catch - not sure why it says test set - changed to training

File: experiments/interpretation/dutchf3_patch/local/train.py (hunk 4 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/train.py:243]: sure. I actually thought this was cleaner.... but just tied it back up to log_validation_results method
